File: analytics/src/api/feedbackAnalytics.ts
import { get } from './client'

// Espelha DTOs em
// src/EfsAiHub.Core.Abstractions/Observability/FeedbackAnalytics.cs.
// Endpoints em
// src/EfsAiHub.Host.Api/Controllers/FeedbackAnalyticsController.cs.

export type Sentiment = 1 | -1

export interface FeedbackTopMessage {
  messageId: string
  feedbackCount: number
  positives: number
  negatives: number
  /** Primeiros ~200 chars de chat_messages.Content. Null se mensagem deletada. */
  messagePreview?: string | null
  conversationId?: string | null
}

export interface FeedbackOverview {
  projectId: string
  periodFrom: string
  periodTo: string
  total: number
  positives: number
  negatives: number
  /** Positives / (Positives + Negatives). 0..1. */
  satisfactionRate: number
  distinctMessages: number
  distinctUsers: number
  withCommentCount: number
  topMessages: FeedbackTopMessage[]
}

export interface FeedbackTimeseriesBucket {
  bucket: string
  count: number
  positives: number
  negatives: number
}

export interface FeedbackRecentRow {
  feedbackId: string
  messageId: string
  conversationId?: string | null
  executionId?: string | null
  /** -1 (dislike) ou +1 (like). */
  sentiment: number
  comment?: string | null
  userId?: string | null
  messagePreview?: string | null
  createdAt: string
}

export interface FeedbackRecentResponse {
  items: FeedbackRecentRow[]
  total: number
  page: number
  pageSize: number
}

export type FeedbackGranularity = 'day' | 'hour'

function buildQs(projectId: string, from?: string, to?: string, extras?: Record<string, string>): string {
  const qs = new URLSearchParams()
  qs.set('projectId', projectId)
  if (from) qs.set('from', from)
  if (to) qs.set('to', to)
  if (extras) for (const [k, v] of Object.entries(extras)) qs.set(k, v)
  return `?${qs.toString()}`
}

export function getFeedbackSummary(
  projectId: string,
  from?: string,
  to?: string,
  init?: RequestInit,
): Promise<FeedbackOverview> {
  return get<FeedbackOverview>(`/analytics/feedback/summary${buildQs(projectId, from, to)}`, init)
}

export function getFeedbackTimeseries(
  projectId: string,
  groupBy: FeedbackGranularity = 'day',
  from?: string,
  to?: string,
  init?: RequestInit,
): Promise<FeedbackTimeseriesBucket[]> {
  return get<FeedbackTimeseriesBucket[]>(
    `/analytics/feedback/timeseries${buildQs(projectId, from, to, { groupBy })}`,
    init,
  )
}

export interface FeedbackRecentFilters {
  /** null | undefined = todos; 1 = likes; -1 = dislikes. */
  sentiment?: Sentiment | null
  page?: number
  pageSize?: number
}

export function getFeedbackRecent(
  projectId: string,
  from?: string,
  to?: string,
  filters: FeedbackRecentFilters = {},
  init?: RequestInit,
): Promise<FeedbackRecentResponse> {
  const extras: Record<string, string> = {}
  if (filters.sentiment === 1 || filters.sentiment === -1) {
    extras.sentiment = String(filters.sentiment)
  }
  if (filters.page) extras.page = String(filters.page)
  if (filters.pageSize) extras.pageSize = String(filters.pageSize)
  return get<FeedbackRecentResponse>(
    `/analytics/feedback/recent${buildQs(projectId, from, to, extras)}`,
    init,
  )
}

File: src/EfsAiHub.Core.Abstractions/Observability/FeedbackAnalytics.cs
namespace EfsAiHub.Core.Abstractions.Observability;

/// <summary>
/// Agregados de feedback do usuário (<c>aihub.message_feedbacks</c>) por
/// projeto. Sentiment é binário ±1 (constraint do schema): +1 = like,
/// -1 = dislike. Comment opcional.
/// </summary>
public sealed class FeedbackOverview
{
    public required string ProjectId { get; init; }
    public DateTime PeriodFrom { get; init; }
    public DateTime PeriodTo { get; init; }

    public long Total { get; init; }
    public long Positives { get; init; }
    public long Negatives { get; init; }
    /// <summary>Positives / (Positives + Negatives). 0 quando ainda não houve feedback.</summary>
    public double SatisfactionRate { get; init; }

    public int DistinctMessages { get; init; }
    public int DistinctUsers { get; init; }
    public long WithCommentCount { get; init; }

    /// <summary>Top 10 mensagens com mais feedback no período (preview da fala incluso).</summary>
    public required IReadOnlyList<FeedbackTopMessageRow> TopMessages { get; init; }
}

public sealed class FeedbackTopMessageRow
{
    public required string MessageId { get; init; }
    public long FeedbackCount { get; init; }
    public long Positives { get; init; }
    public long Negatives { get; init; }
    /// <summary>Primeiros ~200 chars do <c>chat_messages.Content</c>. Null se mensagem foi deletada.</summary>
    public string? MessagePreview { get; init; }
    public string? ConversationId { get; init; }
}

public sealed class FeedbackTimeseriesBucket
{
    public DateTime Bucket { get; init; }
    public long Count { get; init; }
    public long Positives { get; init; }
    public long Negatives { get; init; }
}

/// <summary>
/// Listagem paginada de feedbacks recentes. Inclui preview da mensagem
/// avaliada (JOIN com <c>chat_messages</c>) — é o que faz a tela útil pra
/// análise qualitativa ("o que levou thumbs down?").
/// </summary>
public sealed class FeedbackRecentResponse
{
    public required IReadOnlyList<FeedbackRecentRow> Items { get; init; }
    public long Total { get; init; }
    public int Page { get; init; }
    public int PageSize { get; init; }
}

public sealed class FeedbackRecentRow
{
    public required string FeedbackId { get; init; }
    public required string MessageId { get; init; }
    public string? ConversationId { get; init; }
    public string? ExecutionId { get; init; }
    /// <summary>-1 (dislike) ou +1 (like). Schema garante via CHECK constraint.</summary>
    public int Sentiment { get; init; }
    public string? Comment { get; init; }
    public string? UserId { get; init; }
    /// <summary>Primeiros ~300 chars do <c>chat_messages.Content</c>. Null se mensagem foi deletada.</summary>
    public string? MessagePreview { get; init; }
    public DateTime CreatedAt { get; init; }
}
